src/Microsoft.OpenApi/Services/OpenApiComparerBase.cs

```diff
@@ -138,6 +166,125 @@ internal void Compare<TEnum>(Enum source, Enum target, ComparisonContext compari
             }
         }
 
+        /// <summary>
+        /// Compares <see cref="OpenApiReference"/> object.
+        /// </summary>
+        /// <param name="sourceReference">The source.</param>
+        /// <param name="targetReference">The target.</param>
+        /// <param name="comparisonContext">The context under which to compare the objects.</param>
+        internal void Compare<TReference>(
+            OpenApiReference sourceReference,
+            OpenApiReference targetReference,
+            ComparisonContext comparisonContext)
+        {
+            if (sourceReference == null && targetReference == null)
+            {
+                return;
+            }
+
+            if (sourceReference == null || targetReference == null)
+            {
+                comparisonContext.AddOpenApiDifference(
+                    new OpenApiDifference
+                    {
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Update,
+                        SourceValue = sourceReference,
+                        TargetValue = targetReference,
+                        OpenApiComparedElementType = typeof(OpenApiReference),
+                        Pointer = comparisonContext.PathString
+                    });
+
+                return;
+            }
+
+            if (sourceReference.Id != targetReference.Id || sourceReference.Type != targetReference.Type)
+            {
+                WalkAndAddOpenApiDifference(
+                    comparisonContext,
+                    OpenApiConstants.DollarRef,
+                    new OpenApiDifference
+                    {
+                        OpenApiDifferenceOperation = OpenApiDifferenceOperation.Update,
+                        SourceValue = sourceReference,
+                        TargetValue = targetReference,
+                        OpenApiComparedElementType = typeof(OpenApiReference)
+                    });
+
+                return;
+            }
+
+            var source = (TReference) comparisonContext.SourceDocument.ResolveReference(
+                sourceReference);
+
+            var target = (TReference) comparisonContext.TargetDocument.ResolveReference(
+                targetReference);
+
+            comparisonContext
+                .GetComparer<TReference>()
+                .Compare(source, target, comparisonContext);
```

Comment:
What's the example case of this? If the references are indeed pointing to the same object, wouldn't these two objects always be the same?

Reply:
if Id are same we can't assume that the referenced object will be same too.

---
In reply to: [223515635](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223515635) [](ancestors = 223515635)

Reply:
That could be true in the C# sense, but from the way we use the Reference object, how would that be possible? If the ID is the same, doesn't that mean this refers to exactly the same object. In fact, the reference object will only be serialized as $ref (or plain string in case of tag) with id only.

---
In reply to: [223858079](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223858079) [](ancestors = 223858079,223515635)

Reply:
Chatted with Shweta. We will do a deep comparison here, so that it's useful for API consumer to see what exactly has changed to an API. Same principle applies to the Security Scheme in Security Requirement.

---
In reply to: [223899238](https://github.com/Microsoft/OpenAPI.NET/pull/328#discussion_r223899238) [](ancestors = 223899238,223858079,223515635)